# Data Cleaning Notebook - Marketing Attribution ROI Dashboard

## Stage 1: Data Quality Assessment
This section evaluates data quality by examining:
- Data types
- Missing values
- Duplicate records
- Data consistency


In [1]:
import pandas as pd
import numpy as np

In [3]:
# Load datasets
campaigns = pd.read_csv("../data/raw/campaigns.csv")
customers = pd.read_csv("../data/raw/customers.csv")
events = pd.read_csv("../data/raw/events.csv")
products = pd.read_csv("../data/raw/products.csv")
transactions = pd.read_csv("../data/raw/transactions.csv")

In [5]:
# Checking for missing values and data shape in each dataset
for df_name, df in {
    "campaigns": campaigns,
    "customers": customers,
    "events": events,
    "products": products,
    "transactions": transactions
}.items():
    print(df_name)
    print("Shape:", df.shape)
    print("Missing Values:", df.isnull().sum().sum())
    print("-" * 40)

campaigns
Shape: (50, 7)
Missing Values: 0
----------------------------------------
customers
Shape: (100000, 7)
Missing Values: 0
----------------------------------------
events
Shape: (2000000, 12)
Missing Values: 240671
----------------------------------------
products
Shape: (2000, 6)
Missing Values: 0
----------------------------------------
transactions
Shape: (103127, 9)
Missing Values: 20898
----------------------------------------


In [ ]:
# Getting summary statistics for each dataset
for name, df in datasets.items():
    print(f"\n{name.upper()} Summary Statistics")
    display(df.describe(include='all'))


CAMPAIGNS Summary Statistics


,campaign_id,channel,objective,start_date,end_date,target_segment,expected_uplift
count,50.00000,50,50,50,50,50,50.000000
unique,NaN,5,4,49,48,5,NaN
top,NaN,Paid Search,Reactivation,2022-07-25,2022-09-29,New Customers,NaN
freq,NaN,11,15,2,2,12,NaN
mean,25.50000,NaN,NaN,NaN,NaN,NaN,0.086660
std,14.57738,NaN,NaN,NaN,NaN,NaN,0.035854
min,1.00000,NaN,NaN,NaN,NaN,NaN,0.022000
25%,13.25000,NaN,NaN,NaN,NaN,NaN,0.057250
50%,25.50000,NaN,NaN,NaN,NaN,NaN,0.092500
75%,37.75000,NaN,NaN,NaN,NaN,NaN,0.113750



CUSTOMERS Summary Statistics


,customer_id,signup_date,country,age,gender,loyalty_tier,acquisition_channel
count,100000.000000,100000,100000,100000.000000,100000,100000,100000
unique,NaN,1095,7,NaN,3,4,5
top,NaN,2022-11-04,US,NaN,Male,Bronze,Organic
freq,NaN,124,34931,NaN,48054,60276,30200
mean,50000.500000,NaN,NaN,35.224370,NaN,NaN,NaN
std,28867.657797,NaN,NaN,9.595792,NaN,NaN,NaN
min,1.000000,NaN,NaN,18.000000,NaN,NaN,NaN
25%,25000.750000,NaN,NaN,28.000000,NaN,NaN,NaN
50%,50000.500000,NaN,NaN,35.000000,NaN,NaN,NaN
75%,75000.250000,NaN,NaN,42.000000,NaN,NaN,NaN



EVENTS Summary Statistics


,event_id,timestamp,customer_id,session_id,event_type,product_id,device_type,traffic_source,campaign_id,page_category,session_duration_sec,experiment_group
count,2.000000e+06,2000000,2.000000e+06,2.000000e+06,2000000,1.799629e+06,1959700,2000000,2.000000e+06,2000000,2.000000e+06,2000000
unique,NaN,1977065,NaN,NaN,5,NaN,3,10,NaN,5,NaN,3
top,NaN,2023-09-18 07:00:22,NaN,NaN,view,NaN,mobile,Organic,NaN,PLP,NaN,Control
freq,NaN,4,NaN,NaN,1043573,NaN,1176146,776758,NaN,601284,NaN,1198404
mean,1.000000e+06,NaN,5.001209e+04,3.330514e+05,NaN,1.000100e+03,NaN,NaN,1.276203e+01,NaN,1.306426e+02,NaN
std,5.773504e+05,NaN,2.887960e+04,1.924715e+05,NaN,5.776716e+02,NaN,NaN,1.634052e+01,NaN,1.590376e+02,NaN
min,1.000000e+00,NaN,1.000000e+00,1.000000e+00,NaN,1.000000e+00,NaN,NaN,0.000000e+00,NaN,1.000000e-01,NaN
25%,5.000008e+05,NaN,2.501000e+04,1.662308e+05,NaN,5.000000e+02,NaN,NaN,0.000000e+00,NaN,5.810000e+01,NaN
50%,1.000000e+06,NaN,5.000900e+04,3.328190e+05,NaN,1.000000e+03,NaN,NaN,0.000000e+00,NaN,1.017000e+02,NaN
75%,1.500000e+06,NaN,7.503800e+04,4.997712e+05,NaN,1.501000e+03,NaN,NaN,2.600000e+01,NaN,1.640000e+02,NaN



PRODUCTS Summary Statistics


,product_id,category,brand,base_price,launch_date,is_premium
count,2000.000000,2000,2000,2000.000000,2000,2000.000000
unique,NaN,6,100,NaN,901,NaN
top,NaN,Electronics,Brand_7,NaN,2021-01-18,NaN
freq,NaN,455,32,NaN,7,NaN
mean,1000.500000,NaN,NaN,72.166650,NaN,0.500000
std,577.494589,NaN,NaN,52.315792,NaN,0.500125
min,1.000000,NaN,NaN,5.110000,NaN,0.000000
25%,500.750000,NaN,NaN,32.505000,NaN,0.000000
50%,1000.500000,NaN,NaN,61.245000,NaN,0.500000
75%,1500.250000,NaN,NaN,97.955000,NaN,1.000000



TRANSACTIONS Summary Statistics


,transaction_id,timestamp,customer_id,product_id,quantity,discount_applied,gross_revenue,campaign_id,refund_flag
count,103127.000000,103127,103127.000000,92678.00000,103127.000000,103127.000000,92678.000000,103127.000000,103127.000000
unique,NaN,103074,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,NaN,2022-11-06 18:59:39,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,NaN,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,51564.000000,NaN,49946.662707,999.63139,1.381539,0.041379,90.355493,19.945252,0.029372
std,29770.344943,NaN,28841.644812,576.19187,0.748176,0.060286,100.684683,16.379672,0.168846
min,1.000000,NaN,2.000000,1.00000,1.000000,0.000000,-783.040000,0.000000,0.000000
25%,25782.500000,NaN,25081.000000,498.00000,1.000000,0.000000,34.240000,4.000000,0.000000
50%,51564.000000,NaN,49957.000000,1002.00000,1.000000,0.000000,68.000000,18.000000,0.000000
75%,77345.500000,NaN,74870.500000,1498.00000,1.000000,0.050000,117.617500,34.000000,0.000000


In [ ]:
for name, df in datasets.items():
    duplicates = df.duplicated().sum()
    print(f"{name}: {duplicates} duplicate rows")

campaigns: 0 duplicate rows
customers: 0 duplicate rows
events: 0 duplicate rows
products: 0 duplicate rows
transactions: 0 duplicate rows


In [ ]:
for name, df in datasets.items():
    print(f"\n{name.upper()} Missing Values")
    print(df.isnull().sum())
    print("-" * 50)


CAMPAIGNS Missing Values
campaign_id        0
channel            0
objective          0
start_date         0
end_date           0
target_segment     0
expected_uplift    0
dtype: int64
--------------------------------------------------

CUSTOMERS Missing Values
customer_id            0
signup_date            0
country                0
age                    0
gender                 0
loyalty_tier           0
acquisition_channel    0
dtype: int64
--------------------------------------------------

EVENTS Missing Values
event_id                     0
timestamp                    0
customer_id                  0
session_id                   0
event_type                   0
product_id              200371
device_type              40300
traffic_source               0
campaign_id                  0
page_category                0
session_duration_sec         0
experiment_group             0
dtype: int64
--------------------------------------------------

PRODUCTS Missing Values
product_id    

## Data Profiling Summary

Five datasets were analyzed:

| Dataset | Rows | Columns | Missing Values |
|----------|---------:|---------:|---------:|
| Campaigns | 50 | 7 | 0 |
| Customers | 100,000 | 7 | 0 |
| Events | 2,000,000 | 12 | 240,671 |
| Products | 2,000 | 6 | 0 |
| Transactions | 103,127 | 9 | 20,898 |

Initial profiling indicates that the Campaigns, Customers, and Products datasets are complete with no missing values. Missing values are concentrated in the Events and Transactions datasets and require further investigation.

In [ ]:
# checking structual summary of each dataset
datasets = {
    "campaigns": campaigns,
    "customers": customers,
    "events": events,
    "products": products,
    "transactions": transactions
}

for name, df in datasets.items():
    print(f"\n{name.upper()}")
    print(df.info())


CAMPAIGNS
<class 'pandas.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   campaign_id      50 non-null     int64  
 1   channel          50 non-null     str    
 2   objective        50 non-null     str    
 3   start_date       50 non-null     str    
 4   end_date         50 non-null     str    
 5   target_segment   50 non-null     str    
 6   expected_uplift  50 non-null     float64
dtypes: float64(1), int64(1), str(5)
memory usage: 2.9 KB
None

CUSTOMERS
<class 'pandas.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype
---  ------               --------------   -----
 0   customer_id          100000 non-null  int64
 1   signup_date          100000 non-null  str  
 2   country              100000 non-null  str  
 3   age                  100000 non-null  int64
 4   gender       

In [ ]:
for name, df in datasets.items():
    print(f"\n{name.upper()}")
    print(df.dtypes)
    print("-" * 60)


CAMPAIGNS
campaign_id          int64
channel                str
objective              str
start_date             str
end_date               str
target_segment         str
expected_uplift    float64
dtype: object
------------------------------------------------------------

CUSTOMERS
customer_id            int64
signup_date              str
country                  str
age                    int64
gender                   str
loyalty_tier             str
acquisition_channel      str
dtype: object
------------------------------------------------------------

EVENTS
event_id                  int64
timestamp                   str
customer_id               int64
session_id                int64
event_type                  str
product_id              float64
device_type                 str
traffic_source              str
campaign_id               int64
page_category               str
session_duration_sec    float64
experiment_group            str
dtype: object
------------------------------

## Data Type Assessment Summary

- All date-related columns are currently stored as strings and will require conversion to datetime format for time-based analysis.
- Product ID is stored as float in Events and Transactions due to missing values, while Products table uses integer format.
- All primary and foreign keys are correctly typed as integers except where missing values exist.
- Boolean fields such as `is_premium` require conversion to boolean type for better interpretability.

## Data Profiling Summary

Five datasets were analyzed as part of the Marketing Attribution dataset:

| Dataset      | Rows      | Columns | Missing Values |
|--------------|----------:|--------:|---------------:|
| Campaigns    | 50        | 7       | 0              |
| Customers    | 100,000   | 7       | 0              |
| Events       | 2,000,000 | 12      | 240,671        |
| Products     | 2,000     | 6       | 0              |
| Transactions | 103,127   | 9       | 20,898         |

---

### 1. Data Quality Overview

- Campaigns, Customers, and Products datasets are complete with no missing values or duplicates.
- Events and Transactions datasets contain missing values that require further investigation.
- No duplicate records were identified across all datasets.

---

### 2. Key Data Issues Identified

- **Events dataset**
  - Missing values in `product_id` and `device_type`
  - Largest dataset with 2,000,000 records

- **Transactions dataset**
  - Missing values in `product_id` and `gross_revenue`
  - Requires validation to distinguish incomplete or invalid transactions

---

### 3. Data Type Observations

- All date-related columns are currently stored as string types and will require conversion to datetime format.
- Product ID is inconsistently represented due to missing values in Events and Transactions.
- Boolean field (`is_premium`) is stored as integer and requires type conversion.

---

### 4. Analytical Implications

- Time-based analysis cannot be performed until datetime conversion is completed.
- Events dataset requires careful handling of missing product-level interactions.
- Transactions dataset requires cleaning rules to handle missing revenue and product linkage.
- Overall dataset structure is consistent and suitable for cleaning and transformation phase
 

## Stage 2: Data Cleaning and Preparation

### Overview
This stage focuses on preparing the raw marketing attribution datasets for analysis. The goal is to ensure data quality, consistency, and analytical readiness before performing customer journey and attribution modeling.

### What is being evaluated in this stage?

In this stage, the following data quality dimensions are assessed and addressed:

- **Missing Data Handling**
  - Identify and treat missing values in Events and Transactions datasets
  - Determine whether to impute, retain, or flag missing records

- **Data Type Standardization**
  - Convert date fields from string format to datetime format
  - Ensure numeric and categorical fields are correctly typed

- **Data Consistency**
  - Align product_id formats across datasets
  - Standardize boolean and categorical fields

- **Data Integrity Checks**
  - Validate transactional completeness
  - Ensure event and customer relationships remain intact

- **Feature Readiness**
  - Prepare datasets for customer journey and attribution analysis
  - Create clean, analysis-ready datasets

### Outcome of this Stage
At the end of this stage, all datasets will be cleaned, standardized, and ready for:
- Customer journey mapping
- Marketing attribution modeling
- KPI and ROI analysis

In [ ]:
# Handle missing values in 'device_type' column, replacing 40,300 rows out of N2m rows NaN with 'Unknown'
events["device_type"] = events["device_type"].fillna("Unknown")

In [ ]:
# Handle missing values in 'product_id' column, replacing 20,000 rows out of N2m rows NaN with 'No_Product'
events["product_id"] = events["product_id"].fillna("No_Product")

In [ ]:
# Convert 'timestamp' column to datetime format for accurate time-based analysis
events["timestamp"] = pd.to_datetime(events["timestamp"])

In [ ]:
# Create a new column 'transaction_status' in the transactions dataset to indicate whether the transaction is valid or invalid based on the presence of 'gross_revenue'
transactions["transaction_status"] = transactions["gross_revenue"].apply(
    lambda x: "valid" if pd.notnull(x) else "invalid"
)

### Investigate transaction quality
The transactions dataset shows that missing product_id and missing gross_revenue is 10, 499 rows which is correspondingly the same.
Thus, this definition:
Valid transactions:
product_id present
gross_revenue present

Invalid/incomplete transactions:
missing product_id OR revenue

In [6]:
# Export cleaned datasets for downstream analysis
import os
os.makedirs("../data/processed", exist_ok=True)

events.to_csv("../data/processed/events_cleaned.csv", index=False)
transactions.to_csv("../data/processed/transactions_cleaned.csv", index=False)
customers.to_csv("../data/processed/customers_cleaned.csv", index=False)
campaigns.to_csv("../data/processed/campaigns_cleaned.csv", index=False)
products.to_csv("../data/processed/products_cleaned.csv", index=False)